Mathlib/Algebra/Group/Hom/CompTypeclasses.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+
+import Mathlib.Logic.Function.CompTypeclasses
+import Mathlib.Algebra.Group.Hom.Defs
+
+/-!
+# Propositional typeclasses on several monoid homs
+
+This file contains typeclasses used in the definition of equivariant maps,
+in the spirit what was initially developed by Frédéric Dupuis and Heather Macbeth for linear maps.
+However, we do not expect that all maps should be guessed automatically,
+as it happens for linear maps.
+
+If `φ`, `ψ`… are monoid homs and `M`, `N`… are monoids, we define two classes:
+* `MonoidHom.CompTriple φ ψ χ`, which expresses that `ψ.comp φ = χ`
+* `MonoidHom.CompTriple.IsId φ`, which expresses that `φ = id`
+
+Some basic lemmas are proved:
+* `MonoidHom.CompTriple.comp` asserts `MonoidHom.CompTriple φ ψ (ψ.comp φ)`
+* `MonoidHom.CompTriple.id_comp` asserts `MonoidHom.CompTriple φ ψ ψ`
+  in the presence of `MonoidHom.IsId φ`
+* its variant `MonoidHom.CompTriple.comp_id`
+
+It also introduces instances:
+* `MonoidHom.CompTriple.IsId (MonoidHom.id M)`
+* `MonoidHom.CompTriple φ ψ χ`, which expresses that `ψ.comp φ = χ`
+
+TODO :
+* align with RingHomCompTriple
+* probably rename MonoidHom.CompTriple as MonoidHomCompTriple
+(or, on the opposite, rename RingHomCompTriple as RingHom.CompTriple)
+* does one need AddHom.CompTriple ?
+
+-/
+
+
+section MonoidHomCompTriple
+
+namespace MonoidHom
+
+/-- Class of composing triples -/
+class CompTriple  {M N P : Type*} [Monoid M] [Monoid N] [Monoid P]
+  (φ : M →* N) (ψ : N →* P) (χ : outParam (M →* P)) : Prop where
+  /-- The maps form a commuting triangle -/
+  comp_eq : ψ.comp φ = χ
+
+attribute [simp] CompTriple.comp_eq
+
+namespace CompTriple
+
+variable {M' : Type*} [Monoid M']
+variable {M N P : Type*} [Monoid M] [Monoid N] [Monoid P]
+
+/-- Class of Id maps -/
+class IsId (σ : M →* M) : Prop where
+  eq_id : σ = MonoidHom.id M
+
+instance {M : Type*} [Monoid M] : IsId (MonoidHom.id M) where
+  eq_id := rfl
+
+instance {σ : M →* M} [IsId σ] : CompTriple.IsId σ where
+  eq_id := by simp only [IsId.eq_id, MonoidHom.coe_mk, OneHom.coe_mk]
+
+instance {φ : M →* N} {ψ : N  →* P} {χ : M →* P} [κ : CompTriple φ ψ χ] :
+    _root_.CompTriple φ ψ χ where
+  comp_eq := by rw [← MonoidHom.coe_comp, κ.comp_eq]
+
+lemma comp {φ : M →* N} {ψ : N →* P} :
+    CompTriple φ ψ (ψ.comp φ) where
```

Comment:
Yes, this is because I felt (by reading the documentation of the LinearMap file) that adding these instances would be a burden to the compiler.

Reply:
I had a PR #11591 restricted to these two comptriple files, and the final lint didn't work. So I revert to the initial version.
There is `comp` which requires a CompTriple instance, and `comp'` where it can be inserted explicitly. On the other hand, Lean4 allows the user to provide values to optional/automatic parameters, so maybe this `comp'` trick (that came from my Lean3 version) is not necessary anymore.